core/src/main/java/org/apache/iceberg/avro/AvroIterable.java

```diff
@@ -78,7 +79,8 @@ public CloseableIterator<D> iterator() {
     if (start != null) {
       if (reader instanceof SupportsRowPosition) {
         ((SupportsRowPosition) reader)
-            .setRowPositionSupplier(() -> AvroIO.findStartingRowPos(file::newStream, start));
+            .setRowPositionSupplier(
+                Suppliers.memoize(() -> AvroIO.findStartingRowPos(file::newStream, start)));
```

Comment:
Previously, this was being done in the `StructValueReader`. If the struct [reader inserted a `PositionReader`](https://github.com/apache/iceberg/pull/9366/files#diff-08597bac062ea934a49ebeb66da72f459dea1da642e1df2cdd2c8ca6884fed4dL626-L631), it would also rewrite the position supplier.

That was a lot of complication for the value reader and didn't work in all cases (for example, if two structs had position columns) so I moved the memoization here. It's simpler that way.